File: src/main/java/org/dreamexposure/ticketbird/objects/guild/Ticket.java
package org.dreamexposure.ticketbird.objects.guild;

import discord4j.common.util.Snowflake;

public class Ticket {
    private final Snowflake guildId;
    private final int number;
    private String project;
    private Snowflake creator;
    private Snowflake channel;
    private Snowflake category;

    private long lastActivity;

    public Ticket(Snowflake _guildId, int _number) {
        guildId = _guildId;
        number = _number;
        project = "N/a";
        lastActivity = System.currentTimeMillis();

        creator = null;
    }

    //Getters
    public Snowflake getGuildId() {
        return guildId;
    }

    public int getNumber() {
        return number;
    }

    public String getProject() {
        return project;
    }

    public Snowflake getCreator() {
        return creator;
    }

    public Snowflake getChannel() {
        return channel;
    }

    public Snowflake getCategory() {
        return category;
    }

    public long getLastActivity() {
        return lastActivity;
    }

    //Setters
    public void setProject(String _project) {
        project = _project;
    }

    public void setCreator(Snowflake _creator) {
        creator = _creator;
    }

    public void setChannel(Snowflake _channel) {
        channel = _channel;
    }

    public void setCategory(Snowflake _category) {
        category = _category;
    }

    public void setLastActivity(long _activity) {
        lastActivity = _activity;
    }
}

File: src/main/kotlin/org/dreamexposure/ticketbird/object/Ticket.kt
package org.dreamexposure.ticketbird.`object`

import discord4j.common.util.Snowflake
import org.dreamexposure.ticketbird.database.TicketData
import java.time.Instant

data class Ticket(
    val guildId: Snowflake,

    val number: Int,

    var project: String = "N/a",

    val creator: Snowflake,

    val participants: List<Snowflake> = listOf(),

    val channel: Snowflake,

    var category: Snowflake,

    var lastActivity: Instant = Instant.now(),

    var transcriptSha256: String? = null,
    var attachmentsSha256: String? = null,
) {
    constructor(data: TicketData): this(
        guildId = Snowflake.of(data.guildId),
        number = data.number,
        project = data.project,
        creator = Snowflake.of(data.creator),
        participants = data.participants
            ?.split(",")
            ?.filter(String::isNotBlank)
            ?.map(Snowflake::of)
            ?.toList() ?: listOf(),
        channel = Snowflake.of(data.channel),
        category = Snowflake.of(data.category),
        lastActivity = Instant.ofEpochMilli(data.lastActivity),
        transcriptSha256 = data.transcriptSha256,
        attachmentsSha256 = data.attachmentsSha256,
    )
}
